devices/src/virtio/vsock/csm/connection.rs

```diff
@@ -0,0 +1,632 @@
+                // connection.
+                pkt.set_op(uapi::VSOCK_OP_RST);
+            }
+        };
+
+        self.rx_cnt += Wrapping(pkt.len());
+        self.last_fwd_cnt_to_peer = self.fwd_cnt;
+
+        Ok(())
+    }
+
+    /// Deliver a guest-generated packet to this connection.
+    ///
+    /// This forwards the data in RW packets to the host stream, and absorbs control packets,
+    /// using them to manage the internal connection state.
+    ///
+    /// Returns:
+    /// always `Ok(())`: the packet has been consumed;
+    ///
+    fn send_pkt(&mut self, pkt: &VsockPacket) -> VsockResult<()> {
+        // Update the peer credit information.
+        self.peer_buf_alloc = pkt.buf_alloc();
+        self.peer_fwd_cnt = Wrapping(pkt.fwd_cnt());
+
+        match self.state {
+            // Most frequent case: this is an established connection that needs to forward some data
+            // to the host stream. Also works for a connection that has begun shutting down, but the
+            // peer still has some data to send.
+            ConnState::Established | ConnState::PeerClosed(_, false)
+                if pkt.op() == uapi::VSOCK_OP_RW =>
+            {
+                if pkt.buf().is_none() {
+                    info!(
+                        "vsock: dropping empty data packet from guest (lp={}, pp={}",
+                        self.local_port, self.peer_port
+                    );
+                    return Ok(());
+                }
+
+                // Unwrapping here is safe, since we just checked `pkt.buf()` above.
+                let buf_slice = &pkt.buf().unwrap()[..(pkt.len() as usize)];
+                if let Err(err) = self.send_bytes(buf_slice) {
+                    // If we can't write to the host stream, that's an unrecoverable error, so we'll
+                    // terminate this connection.
+                    info!(
+                        "vsock: error writing to local stream (lp={}, pp={}): {:?}",
+                        self.local_port, self.peer_port, err
+                    );
+                    self.kill();
+                    return Ok(());
+                }
+
+                // We might've just consumed some data. If that's the case, we might need to update
+                // the peer on our buffer space situation, so that it can keep sending data packets
+                // our way.
+                if self.peer_needs_credit_update() {
+                    self.pending_rx.insert(PendingRx::CreditUpdate);
+                }
+            }
+
+            // Next up: receing a response / confirmation for a host-initiated connection. We'll
+            // move to an Established state, and pass on the good news through the host stream.
+            ConnState::LocalInit if pkt.op() == uapi::VSOCK_OP_RESPONSE => {
+                self.expiry = None;
+                self.send_bytes(format!("OK {}\n", self.local_port).as_bytes())
```

Comment:
This is probably very very unlikely, but I don't think it's impossible to be in a situation where the next write does not manage to transmit the entire message here. I guess one option is to simply generate an error if we can't write them all in one go.